### PR TITLE
Adds a bunch of player-related functions

### PR DESCRIPTION
**Pull request by ProfessorBoxtrot**

### CODE REVIEW

File: lua/starfall/libs_sh/ply.lua [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+local function CanUsePlyCore( Play )
+    return Play:IsAdmin() or Play:IsSuperAdmin()
+end
+
+
+--- Gets the slow walking speed of a player. This is when you're using +WALK.
+-- @return Slow-walk speed of the player
+function player_methods:getSlowWalkSpeed()
+    if not CanUsePlyCore(instance.player) then SF.Throw("You must be an admin to use this function!") end
```
Comment (overlordbleck) [inline lua/starfall/libs_sh/ply.lua:30]:
```suggestion
    if not CanUsePlyCore( instance.player ) then SF.Throw( "You must be an admin to use this function!" ) end
```

File: lua/starfall/libs_sh/ply.lua [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+    if ent:IsValid() then
+        return ent
+    else
+        SF.Throw( "Player is not valid.", 3 )
+    end
+end
+
+
+local function CanUsePlyCore( Play )
```
Comment (overlordbleck) [inline lua/starfall/libs_sh/ply.lua:22]:
```suggestion
local function canUsePlyCore( Play )
```

File: lua/starfall/libs_sh/ply.lua [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+end
+
+
+--- Gets the regular walking speed of a player. This is when you're just holding W/A/S/D and not using +WALK.
+-- @return Walk speed of the player
+function player_methods:getWalkSpeed()
+    checktype( self, ply_meta )
+    local ent = getPly( self )
+    
+    return ent:GetWalkSpeed()
```
Comment (overlordbleck) [inline lua/starfall/libs_sh/ply.lua:45]:
```suggestion
    return getPly( self ):GetWalkSpeed()
```
or
```suggestion
    local ply = getPly( self )
    
    return ply:GetWalkSpeed()
```

Thoughts?

Comment (overlordbleck) [reply lua/starfall/libs_sh/ply.lua:45]: This follows all other methods as well

File: lua/starfall/libs_sv/ply.lua [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+
+    if ent:IsValid() then
+        return ent
+    else
+        SF.Throw( "Player is not valid.", 3 )
+    end
+end
+
+
+local function CanUsePlyCore( play )
```
Comment (overlordbleck) [inline lua/starfall/libs_sv/ply.lua:24]:
```suggestion
local function canUsePlyCore( play )
```

File: lua/starfall/libs_sv/ply.lua [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    return play:IsAdmin() or play:IsSuperAdmin()
+end
+
+
+--- Sets the health of a player.
+-- @param number Health
+function player_methods:setHealth( health )
+    if not CanUsePlyCore(instance.player) then SF.Throw("You must be an admin to use this function!") end
```
Comment (overlordbleck) [inline lua/starfall/libs_sv/ply.lua:32]:
```suggestion
    if not CanUsePlyCore( instance.player ) then SF.Throw( "You must be an admin to use this function!" ) end
```

Comment (overlordbleck) [reply lua/starfall/libs_sv/ply.lua:32]: Double-check your spacings!

File: lua/starfall/libs_sv/ply.lua [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+
+
+--- Sets the eye angles of a player.
+-- @param angle EyeAngles
+function player_methods:setEyeAngles( ang )
+    if not CanUsePlyCore(instance.player) then SF.Throw("You must be an admin to use this function!") end
+    
+    checktype( self, ply_meta )
+    local ent = getPly( self )
+    local angle = aunwrap(ang)
```
Comment (overlordbleck) [inline lua/starfall/libs_sv/ply.lua:147]:
```suggestion
    local angle = aunwrap( ang )
```

File: lua/starfall/libs_sv/ply.lua [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+    ent:StripAmmo()
+end
+
+
+--- Sets the ammo of the player. Ammo ID can either be a number or a string. Refer to the keys given from the Player:getAmmo() table for numeric IDs.
+-- @param any AmmoID
+-- @param number Amount
+function player_methods:setAmmo( ammoId, amount )
+    if not CanUsePlyCore(instance.player) then SF.Throw("You must be an admin to use this function!") end
+    if type(ammoId) ~= "string" and type(ammoId) ~= "number" then SF.Throw("Ammo ID must either be a string or a number!") end
```
Comment (overlordbleck) [inline lua/starfall/libs_sv/ply.lua:167]: spacings

File: lua/starfall/libs_sv/ply.lua [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+end
+
+
+--- Sets how much ammo the weapon has in its primary clip. It is possible to exceed the weapon's regular max clip ammount with this function.
+-- @param number Ammo
+function weapon_methods:setClip1( amount )
+    if not CanUsePlyCore(instance.player) then SF.Throw("You must be an admin to use this function!") end
+    
+    checktype( self, wep_meta )
+    local ent = wunwrap(self)
```
Comment (overlordbleck) [inline lua/starfall/libs_sv/ply.lua:195]:
```suggestion
    local ent = wunwrap( self )
```

**Review by overlordbleck — COMMENTED** [review]
Looks great! Just a few style errors and we'll be all good to go!

File: lua/starfall/libs_sh/ply.lua [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    
+    return ent:GetAmmo()
+end
+
+
+--- Gets the name (string ID) of an ammo, given an ID. Refer to the keys given from the Player:getAmmo() table for numeric IDs.
+-- @param number ID
+-- @return The name of the given ammo ID
+function builtins_library.getAmmoName( id )
+    return game.GetAmmoName(id)
```
Comment (overlordbleck) [inline lua/starfall/libs_sh/ply.lua:81]:
```suggestion
    return game.GetAmmoName( id )
```

File: lua/starfall/libs_sh/ply.lua [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+end
+
+
+--- Gets the ID (numerical ID) of an ammo, given its name. Inverse of getAmmoName(id)
+-- @param string Name
+-- @return The ID of the given ammo name
+function builtins_library.getAmmoID( name )
+    return game.GetAmmoID(name)
```
Comment (overlordbleck) [inline lua/starfall/libs_sh/ply.lua:89]:
```suggestion
    return game.GetAmmoID( name )
```

File: lua/starfall/libs_sv/ply.lua [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+
+
+local function getPly( this )
+    local ent = punwrap( this )
+
+    if ent:IsValid() then
+        return ent
+    else
+        SF.Throw( "Player is not valid.", 3 )
+    end
```
Comment (overlordbleck) [inline lua/starfall/libs_sv/ply.lua:20]:
```suggestion
    if ent:IsValid() then
        return ent
    end
    
    SF.Throw( "Player is not valid.", 3 )
```

File: lua/starfall/libs_sv/ply.lua [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+end
+
+
+local function canUsePlyCore( play )
+    return play:IsAdmin() or play:IsSuperAdmin()
+end
```
Comment (overlordbleck) [inline lua/starfall/libs_sv/ply.lua:26]:
```suggestion
local function canUsePlyCore( ply )
    return ply:IsAdmin() or ply:IsSuperAdmin()
end
```

File: lua/starfall/libs_sv/ply.lua [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+    
+    checktype( self, ply_meta )
+    local ply = getPly( self )
+    ply:SetMaxHealth( maxHealth )
+end
+
+
+--- Sets the suit armor of a player.
+-- @param number SuitArmor
+function player_methods:setArmor( armor )
```
Comment (overlordbleck) [inline lua/starfall/libs_sv/ply.lua:53]: Should we check if `armor` is a number?

File: lua/starfall/libs_sv/ply.lua [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+    
+    checktype( self, ply_meta )
+    local ply = getPly( self )
+    ply:SetArmor( armor )
+end
+
+
+--- Sets the jump power of a player. Default value is 200.
+-- @param number JumpPower
+function player_methods:setJumpPower( jumpPower )
```
Comment (overlordbleck) [inline lua/starfall/libs_sv/ply.lua:64]: same here ^

File: lua/starfall/libs_sv/ply.lua [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+end
+
+
+--- Gives ammo to the player. Ammo ID can either be a number or a string. Refer to the keys given from the Player:getAmmo() table for numeric IDs.
+-- @param any AmmoID
+-- @param number Amount
+-- @param bool HidePopup
+function player_methods:giveAmmo( ammoId, amount, hidePopup )
+    if not canUsePlyCore( instance.player ) then SF.Throw( "You must be an admin to use this function!" ) end
+    if type(ammoId) ~= "string" and type(ammoId) ~= "number" then SF.Throw("Ammo ID must either be a string or a number!") end
```
Comment (overlordbleck) [inline lua/starfall/libs_sv/ply.lua:181]:
```suggestion
    if type( ammoId ) ~= "string" and type( ammoId ) ~= "number" then SF.Throw( "Ammo ID must either be a string or a number!" ) end
```

File: lua/starfall/libs_sh/ply.lua [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+
+
+local function getPly( this )
+    local ent = punwrap( this )
+
+    if ent:IsValid() then
+        return ent
+    else
+        SF.Throw( "Player is not valid.", 3 )
+    end
```
Comment (overlordbleck) [inline lua/starfall/libs_sh/ply.lua:18]:
```suggestion
    if ent:IsValid() then
        return ent
    end
    
    SF.Throw( "Player is not valid.", 3 )
```

File: lua/starfall/libs_sh/ply.lua [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+end
+
+
+local function canUsePlyCore( play )
+    return play:IsAdmin() or play:IsSuperAdmin()
+end
```
Comment (overlordbleck) [inline lua/starfall/libs_sh/ply.lua:24]:
```suggestion
local function canUsePlyCore( ply )
    return ply:IsAdmin() or ply:IsSuperAdmin()
end
```

**Review by overlordbleck — CHANGES_REQUESTED** [review]
Make sure you're checking the input values to see if its what you want via the `checkluatype` function

File: lua/starfall/libs_sv/ply.lua (hunk 13 above)
Comment (ProfessorBoxtrot) [reply lua/starfall/libs_sv/ply.lua:53]: From my testing, starfall seems to do automatic typechecking, shouldn't be required.

Comment (overlordbleck) [reply lua/starfall/libs_sv/ply.lua:53]: I worry it might spam errors in the server console, however.

Comment (ProfessorBoxtrot) [reply lua/starfall/libs_sv/ply.lua:53]: Does that actually happen? When I tested that, it just gave me a starfall error saying there was a type mismatch.

File: lua/starfall/libs_sh/ply.lua [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+local function canUsePlyCore( ply )
+    return ply:IsAdmin() or ply:IsSuperAdmin()
+end
+
+
+--- Gets the slow walking speed of a player. This is when you're using +WALK.
+-- @return Slow-walk speed of the player
+function player_methods:getSlowWalkSpeed()
+    checktype( self, ply_meta )
+    local ent = getPly( self )
```
Comment (plally) [inline lua/starfall/libs_sh/ply.lua:31]: `local ply = getPly( self )` would be better
there are a few other functions that also need this change

File: lua/starfall/libs_sv/ply.lua [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+    ply:StripAmmo()
+end
+
+
+--- Sets the ammo of the player. Ammo ID can either be a number or a string. Refer to the keys given from the Player:getAmmo() table for numeric IDs.
+-- @param any AmmoID
+-- @param number Amount
+function player_methods:setAmmo( ammoId, amount )
+    if not canUsePlyCore( instance.player ) then SF.Throw( "You must be an admin to use this function!", 2 ) end
+    if type( ammoId ) ~= "string" and type( ammoId ) ~= "number" then SF.Throw( "Ammo ID must either be a string or a number!", 2 ) end
```
Comment (plally) [inline lua/starfall/libs_sv/ply.lua:189]:
```suggestion
    if type( ammoId ) ~= "string" and type( ammoId ) ~= "number" then 
        SF.Throw( "Ammo ID must either be a string or a number!", 2 ) 
    end
```
this line is starting to be on the long side, writing it as multiple lines like so would be preferable 
```

File: lua/starfall/libs_sv/ply.lua [hunk 20]
```diff
@@ -0,0 +1,241 @@
...
+end
+
+
+--- Gives ammo to the player. Ammo ID can either be a number or a string. Refer to the keys given from the Player:getAmmo() table for numeric IDs.
+-- @param any AmmoID
+-- @param number Amount
+-- @param bool HidePopup
+function player_methods:giveAmmo( ammoId, amount, hidePopup )
+    if not canUsePlyCore( instance.player ) then SF.Throw( "You must be an admin to use this function!", 2 ) end
+    if type( ammoId ) ~= "string" and type( ammoId ) ~= "number" then SF.Throw( "Ammo ID must either be a string or a number!", 2 ) end
```
Comment (plally) [inline lua/starfall/libs_sv/ply.lua:205]: same thing as previous function

**Review by plally — CHANGES_REQUESTED** [review]

File: lua/starfall/libs_sh/ply.lua [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+    if ent:IsValid() then
+        return ent
+    end
+    
+    SF.Throw( "Player is not valid.", 3 )
+end
+
+
+local function canUsePlyCore( ply )
+    return ply:IsAdmin() or ply:IsSuperAdmin()
```
[the remark this answers is not in the document]
Comment (plally) [reply lua/starfall/libs_sh/ply.lua:23]: ply:IsAdmin() does return true for superadmins
```suggestion
    return ply:IsAdmin()
```

**Review by plally — APPROVED** [review]

**Pull request merged**